src/join/composer.jl

```diff
@@ -337,516 +113,124 @@ function compose_inner_table(joiner::DataFrameJoiner,
-    ngroups = idx_r_start - 1
-    right_len = length(right)
-    groups = Vector{Int}(undef, right_len)
-    groups[1:ngroups] = 1:ngroups
-
-    @inbounds for idx_r in idx_r_start:right_len
-        val_r = right[idx_r]
-        i = ismissing(val_r) ? length(dict) : Int(val_r) + offset
-        dict_val = dict[i]
-        if dict_val > 0
-            groups[idx_r] = dict_val
-        else
-            ngroups += 1
-            groups[idx_r] = ngroups
-            dict[i] = ngroups
-        end
+    if isnothing(indicator)
+        src_indicator = nothing
+    else
+        src_indicator = Vector{UInt32}(undef, target_nrow)
+        src_indicator[1:lil] .= 3
+        src_indicator[lil + 1:lil + loil] .= 1
+        src_indicator[lil + loil + 1:target_nrow] .= 2
     end
 
-    @assert ngroups > 0 # we should not get here with 0-length right
-    return _innerjoin_postprocess_int(left, dict, groups, ngroups, right_len, offset, minv, maxv)
-end
-
-function compute_join_indices!(groups::Vector{Int}, ngroups::Int,
-                               starts::Vector, rperm::Vector)
-    @inbounds for gix in groups
-        starts[gix] += 1
-    end
+    cols = Vector{AbstractVector}(undef, ncol(joiner.dfl) + ncol(dfr_noon))
 
-    cumsum!(starts, starts)
+    col_idx = 1
 
-    @inbounds for (i, gix) in enumerate(groups)
-        rperm[starts[gix]] = i
-        starts[gix] -= 1
-    end
-    push!(starts, length(groups))
-    return nothing
-end
+    left_idxs = [columnindex(joiner.dfl, n) for n in joiner.left_on]
+    append!(left_idxs, setdiff(1:ncol(joiner.dfl), left_idxs))
 
-function _innerjoin_postprocess(left::AbstractArray, dict::Dict{T, Int},
-                                groups::Vector{Int}, ngroups::Int, right_len::Int) where {T}
-    starts = zeros(Int, ngroups)
-    rperm = Vector{Int}(undef, right_len)
-
-    left_ixs = Int[]
-    right_ixs = Int[]
-
-    # lower bound assuming we get matches
-    sizehint!(left_ixs, right_len)
-    sizehint!(right_ixs, right_len)
-
-    compute_join_indices!(groups, ngroups, starts, rperm)
-
-    n = 0
-    @inbounds for (idx_l, val_l) in enumerate(left)
-        group_id = get(dict, val_l, -1)
-        if group_id != -1
-            ref_stop = starts[group_id + 1]
-            l = ref_stop - starts[group_id]
-            newn = n + l
-            resize!(left_ixs, newn)
-            for i in n+1:n+l
-                left_ixs[i] = idx_l
-            end
-            resize!(right_ixs, newn)
-            for i in 1:l
-                right_ixs[n + i] = rperm[ref_stop - i + 1]
-            end
-            n = newn
+    if kind == :left
+        @assert loil + lil == target_nrow
+        for col in eachcol(joiner.dfl_on)
+            cols_i = left_idxs[col_idx]
+            cols[cols_i] = similar(col, target_nrow)
+            copyto!(cols[cols_i], view(col, left_ixs))
+            copyto!(cols[cols_i], lil + 1, view(col, leftonly_ixs), 1, loil)
+            col_idx += 1
```

Comment:
@nalimilan - this part is crucial for `leftjoin` performance on pooled vectors. Similar lines using `copyto!` are below for `rightjoin` and `outerjoin`.